client/state/data-layer/wpcom/sites/media/index.js

```diff
@@ -0,0 +1,30 @@
+/**
+ * Internal dependencies
+ */
+import {
+	MEDIA_FILE_UPLOAD_FAILURE,
+	MEDIA_FILE_UPLOAD_SUCCESS,
+	MEDIA_FILE_UPLOADS_ENQUEUE
+} from 'state/action-types';
+import { uploadFile } from 'state/media/actions';
+import { isMediaUploadInProgress, getNextQueuedUpload } from 'state/media/selectors';
+
+export function uploadNext( store ) {
+	const nextQueued = getNextQueuedUpload( store.getState() );
+	if ( nextQueued ) {
+		const { siteId, file } = nextQueued;
+		store.dispatch( uploadFile( siteId, file ) );
+	}
+}
+
+export function maybeUploadFirst( store, action ) {
+	if ( ! isMediaUploadInProgress( store.getState() ) ) {
+		store.dispatch( uploadFile( action.siteId, action.files[ 0 ] ) );
+	}
+}
+
+export default {
+	[ MEDIA_FILE_UPLOADS_ENQUEUE ]: [ maybeUploadFirst ],
+	[ MEDIA_FILE_UPLOAD_FAILURE ]: [ uploadNext ],
```

Comment:
>this would be a great place to inject some other responder such as `announceFailure` which could dispatch a `notice` with a simple message explaining the failure

Indeed, we'll need something like this to recreate the current media UI in Redux. The Flux implementation defines a [`MediaValidationStore`](https://github.com/Automattic/wp-calypso/blob/master/client/lib/media/validation-store.js) to track errors. To recreate this, a reducer following the failure actions could track these errors, though this brings us back to the issue of losing dropped actions. Along these lines, if we did want to continue dropping actions, it would seem the next best thing would be for the network handlers to be responsible for dispatching back into our store the equivalent actions those failures would incur?